cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -439,24 +542,42 @@ public void visitEnd() {
    * @return a new {@link Map} where the key is the parameter name given in the annotation and value is the
    * position of the parameter.
    */
-  private static Map<String, Integer> processParameterAnnotationNode(Map<Integer, AnnotationNode>
+  private static Map<String, Integer> processParameterAnnotationNode(Map<Integer, ParameterDetail>
                                                                        parameterAnnotationNode) {
     Map<String, Integer> parameterAnnotation = new HashMap<>();
-    for (Map.Entry<Integer, AnnotationNode> entry : parameterAnnotationNode.entrySet()) {
+    for (Map.Entry<Integer, ParameterDetail> entry : parameterAnnotationNode.entrySet()) {
       // the AnnotationNode of parameter will be an array of size 2 where 0 is "value" and 1 is the name provided in
       // the annotation.
-      String paraName = (String) entry.getValue().values.get(1);
-      // We expect all parameters to have unique names to ensure that here
-      Preconditions.checkArgument(!parameterAnnotation.containsKey(paraName),
-                                  String.format("A parameter with name %s was already found at position %s." +
-                                                  " Please use unique names.", paraName,
-                                                parameterAnnotation.get(paraName)));
-      // store the parameter position with the unique name given to it
-      parameterAnnotation.put(paraName, entry.getKey());
+      String paraName = (String) entry.getValue().getAnnotationNode().values.get(1);
+      int position = entry.getKey();
+      // if we have already seen a parameter named with this annotation then decide the preferred parameter if possible
+      if (parameterAnnotation.containsKey(paraName)) {
+        position = getPreferredParameter(parameterAnnotationNode.get(parameterAnnotation.get(paraName)),
+                parameterAnnotation.get(paraName), entry.getValue(), entry.getKey());
+
+      }
+      // store the parameter position with this name
+      parameterAnnotation.put(paraName, position);
     }
     return parameterAnnotation;
   }
 
+  private static int getPreferredParameter(ParameterDetail previousDetails, int prevPosition,
+                                           ParameterDetail currentDetails, int currentPosition) {
+    if (Name.class.getName().equals(Type.getType(previousDetails.getAnnotationNode().desc).getClassName()) ||
```

Comment:
It's better to use `Type` to do the comparison, then using the name. The code will be easier to read.